src/System.Net.Sockets/src/System/Net/Sockets/TCPClient.Unix.cs

```diff
@@ -0,0 +1,424 @@
+            set { SetOption(ref _lingerState, ref _lingerStateInitialized, SocketOptionLevel.Socket, SocketOptionName.Linger, value); }
+        }
+
+        private bool NoDelayCore
+        {
+            get { return GetOption(ref _noDelay, ref _noDelayInitialized, SocketOptionLevel.Tcp, SocketOptionName.NoDelay) != 0; }
+            set { SetOption(ref _noDelay, ref _noDelayInitialized, SocketOptionLevel.Tcp, SocketOptionName.NoDelay, value ? 1 : 0); }
+        }
+
+        private bool ExclusiveAddressUseCore
+        {
+            get
+            {
+                return GetOption(ref _exclusiveAddressUse, ref _exclusiveAddressUseInitialized, SocketOptionLevel.Socket, SocketOptionName.ExclusiveAddressUse) != 0;
+            }
+            set
+            {
+                // Unlike the rest of the properties, we code this one explicitly, so as to take advantage of validation done by Socket's property.
+                _exclusiveAddressUse = value ? 1 : 0;
+                _exclusiveAddressUseInitialized = true;
+                if (_clientSocket != null)
+                {
+                    _clientSocket.ExclusiveAddressUse = value; // Use setter explicitly as it does additional validation beyond that done by SetOption
+                }
+            }
+        }
+
+        private T GetOption<T>(ref T location, ref bool initialized, SocketOptionLevel level, SocketOptionName name)
+        {
+            // Used in the getter for each shadow property.  
+
+            // If we already have our client socket set up, just get its value for the option.  
+            // Otherwise, return our shadow property.  If that shadow hasn't yet been initialized, 
+            // first initialize it by creating a temporary socket from which we can obtain a default value.
+
+            if (_clientSocket != null)
+            {
+                return (T)_clientSocket.GetSocketOption(level, name);
+            }
+
+            if (!initialized)
+            {
+                using (Socket s = CreateSocket())
+                {
+                    location = (T)s.GetSocketOption(level, name);
+                }
+                initialized = true;
+            }
+
+            return location;
+        }
+
+        private void SetOption<T>(ref T location, ref bool initialized, SocketOptionLevel level, SocketOptionName name, T value)
+        {
+            // Used in the setter for each shadow property.
+
+            // Store the value into our shadow
+            location = value;
+            initialized = true;
+
+            // Then store the option on to the client socket.  If the client socket isn't yet set, we still want to set
+            // the property onto a socket so we can get validation performed by the underlying system on the option being
+            // set... so, albeit being a bit expensive, we create a temporary socket we can use for validation purposes.
+            Socket s = _clientSocket ?? CreateSocket();
```

Comment:
Should we not do this validation before storing the value into the shadow property? As written it looks like getting and setting options will work differently than it would if the underlying socket was not temporary.